### PR TITLE
8326012: JFR: Event for time to safepoint

### PR DESCRIPTION
**Pull request by D-D-H**
There are now some JFR events related to safepoint. When time-to-safepoint (aka ttsp) is too long, these events could not be very helpful since based on them we cannot know which threads cause it and what those threads are doing.

Users can use `-XX:+SafepointTimeout -XX:SafepointTimeoutDelay=100` to see the threads that don't reach safepoint in time but without stack traces. Using `-XX:+ AbortVMOnSafepointTimeout` can capture the stack traces but it crashes the process, hence it's not sensible to enable the flag in production.

~~This patch adds a new JFR event `EventSafepointTimeout` to record the threads that cause ttsp too long.~~

~~This event includes two fields:~~

~~- safepointId: the relevant safepoint id~~
~~- timeExceeded: the amount of time exceeding `SafepointTimeoutDelay` used by the thread to reach safepoint~~

~~In the current version, this event records the stack of those problematic threads when they finally reach safepoint. Hence, there is a bias, but it's still helpful to deduce the root place.~~

A better implementation is to record a more accurate stack, but this will increase complexity. At the same time, the native stack may also be important for this problem, but it is not currently supported by JFR.

Any input would be greatly appreciated.

Testing: jdk/jdk/jfr

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326012](https://bugs.openjdk.org/browse/JDK-8326012): JFR: Event for time to safepoint (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17888/head:pull/17888` \
`$ git checkout pull/17888`

Update a local copy of the PR: \
`$ git checkout pull/17888` \
`$ git pull https://git.openjdk.org/jdk.git pull/17888/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17888`

View PR using the GUI difftool: \
`$ git pr show -t 17888`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17888.diff">https://git.openjdk.org/jdk/pull/17888.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17888#issuecomment-1948422878)

### CODE REVIEW

**Comment by egahlin** [comment]
Can the event use the settings enabled and threshold instead of XX:+SafepointTimeout -XX:SafepointTimeoutDelay=x? Then the event can be configured from command line, a .jfc file and remotely over JMX. Perhaps the event can be enabled by default with a high threshold? Would this work, or am I missing something?

**Comment by D-D-H** [comment]
> Can the event use the settings enabled and threshold instead of XX:+SafepointTimeout -XX:SafepointTimeoutDelay=x? Then the event can be configured from command line, a .jfc file and remotely over JMX. Perhaps the event can be enabled by default with a high threshold? Would this work, or am I missing something?

It works.

If implemented in this way, I think `EventTimeToSafepoint` is a more suitable name.

`start_time` of this event is `SafepointTracing::_last_safepoint_begin_time_ns`, and `end_time` is the time when vm_thread found the target thread is not running.

What do you think?

**Comment by D-D-H** [comment]
Convert to draft.

**Comment by D-D-H** [comment]
I improved the implementation and updated the issue title.

This patch introduces a new event `TimeToSafepoint` with the following fields:

- safepointId: like other safepoint-related events

- thread: the target Java thread

- iterations: the number of state check iterations for the thread

The implementation:

- First collect the ttsp information during the thread state check iterations. Note that even if the threshold is 0, the number of events may not be equal to the number of Java threads since the information will not be collected in the first `for` loop (it's not necessary in my opinion). 

- Then emit events at the end of SafepointSynchronize::begin, where the safepoint has been reached and stack traces can be walked safely. There is a bias, but it's still helpful.

File: src/hotspot/share/jfr/support/jfrTimeToSafepoint.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ public:
+  static void on_synchronizing();
+  static void on_thread_not_running(JavaThread* thread, int iterations);
+  static void on_synchronized();
+
+ private:
+  struct Entry {
+    JavaThread* thread;
+    JfrTicks end;
+    int iterations;
```
Comment (D-D-H) [inline src/hotspot/share/jfr/support/jfrTimeToSafepoint.hpp:42]: Maybe we can think about putting them into JfrThreadLocal.

**Comment by D-D-H** [comment]
Any thoughts?

**Comment by egahlin** [comment]
> Any thoughts?

I haven't look at the implementation, mostly interested in the design for now. Should thread-local handshake be included in this event, or is it better done separately, or not at all?

**Comment by D-D-H** [comment]
> Should thread-local handshake be included in this event, or is it better done separately, or not at all?

In my opinion, if we need to record information for thread-local handshake, we'd better do it separately.

We may need more than one event and or different fields than the event that this feature focuses on. Some information that I think we need to record:
- requester thread
- target thread(s)
- the time when the target thread starts processing

As for whether we need to add JFR events for thread-local handshake, I want to first see if there are any problems in real workloads due to slow thread-local handshake.

**Comment by D-D-H** [comment]
Hi @egahlin,

Do you have time to review this patch?

**Comment by D-D-H** [comment]
Gentle ping. Could anyone review this patch?

File: src/hotspot/share/jfr/support/jfrTimeToSafepoint.inline.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+#ifndef SHARE_JFR_SUPPORT_JFRTIMETOSAFEPOINT_INLINE_HPP
+#define SHARE_JFR_SUPPORT_JFRTIMETOSAFEPOINT_INLINE_HPP
+
+#include "jfr/jfrEvents.hpp"
+#include "jfr/recorder/jfrEventSetting.inline.hpp"
+#include "jfr/support/jfrTimeToSafepoint.hpp"
+
+#include "runtime/safepoint.hpp"
+#include "runtime/vmThread.hpp"
```
Comment (stefank) [inline src/hotspot/share/jfr/support/jfrTimeToSafepoint.inline.hpp:32]:
```suggestion
#include "jfr/support/jfrTimeToSafepoint.hpp"

#include "jfr/jfrEvents.hpp"
#include "jfr/recorder/jfrEventSetting.inline.hpp"
#include "runtime/safepoint.hpp"
#include "runtime/vmThread.hpp"
```

File: src/hotspot/share/runtime/safepoint.cpp [hunk 3]
```diff
@@ -71,6 +71,10 @@
 #include "utilities/macros.hpp"
 #include "utilities/systemMemoryBarrier.hpp"
 
+#if INCLUDE_JFR
```
Comment (stefank) [inline src/hotspot/share/runtime/safepoint.cpp:74]: Blankline at 73 should be removed

File: src/hotspot/share/jfr/support/jfrTimeToSafepoint.inline.hpp (hunk 2 above)
Comment (D-D-H) [reply src/hotspot/share/jfr/support/jfrTimeToSafepoint.inline.hpp]: Updated.

File: src/hotspot/share/runtime/safepoint.cpp (hunk 3 above)
Comment (D-D-H) [reply src/hotspot/share/runtime/safepoint.cpp:73]: Removed.

**Comment by D-D-H** [comment]
There are no more feedbacks.

I plan to close it.

**Pull request closed without merging**